cluster-autoscaler/cloudprovider/azure/azure_vms_pool.go

```diff
@@ -102,28 +120,260 @@ func (agentPool *VMsPool) MaxSize() int {
+			// set the count for first manual scale profile to the new target value
+			virtualMachineProfile := *versionedAP.Properties.VirtualMachinesProfile
+			versionedAP.Properties.VirtualMachinesProfile.Scale.Manual[0].Count = to.Int32Ptr(int32(count))
+			requestBody.Properties.VirtualMachinesProfile = &virtualMachineProfile
+			return requestBody, nil
+		}
+		// aks-managed CAS will be using AutoscaleProfile
+		// TODO(wenxuan): support AutoscaleProfile once it become available with the new API release
+	}
+	return armcontainerservice.AgentPool{}, fmt.Errorf("failed to build request body for scale up, agentpool doesn't have valid virtualMachinesProfile")
+}
+
+// scaleUpToCount sets node count for vms agent pool to target value through PUT AP call.
+func (agentPool *VMsPool) scaleUpToCount(count int64) error {
+	agentPool.sizeMutex.Lock()
+	defer agentPool.sizeMutex.Unlock()
+
+	updateCtx, updateCancel := getContextWithCancel()
+	defer updateCancel()
+
+	requestBody, err := agentPool.buildRequestBodyForScaleUp(count)
+	if err != nil {
+		klog.Errorf("Failed to build request body for scale up, error: %s", err)
+		return err
+	}
+
+	klog.V(1).Infof("Scaling up vms pool %s to new target count: %d", agentPool.Name, count)
+	poller, err := agentPool.manager.azClient.agentPoolClient.BeginCreateOrUpdate(
+		updateCtx,
+		agentPool.clusterResourceGroup,
+		agentPool.clusterName,
+		agentPool.Name,
+		requestBody, nil)
+
+	if err != nil {
+		klog.Errorf("Failed to update agentpool %s in cluster %s with error: %v",
+			agentPool.azureRef.Name, agentPool.clusterName, err)
+		return err
+	}
+
+	updateCtx, cancel := getContextWithTimeout(updateRequestContextTimeout)
+	defer cancel()
+
+	if _, err = poller.PollUntilDone(updateCtx, nil); err == nil {
+		// success path
+		klog.Infof("agentPoolClient.BeginCreateOrUpdate for aks cluster %s agentpool %s succeeded", agentPool.clusterName, agentPool.Name)
+		agentPool.curSize = count
+		agentPool.lastSizeRefresh = time.Now()
+		klog.V(6).Infof("setVMsPoolNodeCount: invalidating cache, new size %d", count)
+		agentPool.manager.invalidateCache()
+		return nil
+	}
+
+	klog.Errorf("agentPoolClient.BeginCreateOrUpdate for aks cluster %s agentpool %s failed with error %s",
+		agentPool.clusterName, agentPool.Name, err)
+	return nil
 }
 
 // DeleteNodes extracts the providerIDs from the node spec and
-// delete or deallocate the nodes from the agent pool based on the scale down policy.
+// delete or deallocate the nodes based on the scale down policy of agentpool.
 func (agentPool *VMsPool) DeleteNodes(nodes []*apiv1.Node) error {
-	// TODO(wenxuan): Implement this method
-	return cloudprovider.ErrNotImplemented
+	currentSize, err := agentPool.getVMsPoolSize()
+	if err != nil {
+		return err
+	}
+
+	// if the target size is smaller than the min size, return an error
+	if int(currentSize) <= agentPool.MinSize() {
+		klog.V(3).Infof("min size %d reached, nodes will not be deleted", agentPool.MinSize())
```

Comment:
Most places in core, we log the error from DeleteNodes() https://github.com/kubernetes/autoscaler/blob/3914bdebd1e9453396408f1f99efa4cc5eeaf078/cluster-autoscaler/core/static_autoscaler.go#L806 for example. 

Should we be logging this twice?

Reply:
Log removed